options/options_parser.cc

```diff
@@ -268,70 +269,89 @@ Status RocksDBOptionsParser::Parse(const ConfigOptions& config_options_in,
+    s = fs->NewSequentialFile(file_name, FileOptions(), &seq_file, nullptr);
+    if (!s.ok()) {
+      return s;
     }
-    if (IsSection(line)) {
-      s = EndSection(config_options, section, title, argument, opt_map);
-      opt_map.clear();
-      if (!s.ok()) {
-        return s;
+
+    LineFileReader lf_reader(
+        std::move(seq_file), file_name, config_options.file_readahead_size,
+        nullptr, std::vector<std::shared_ptr<EventListener>>{}, nullptr, retry);
+
+    OptionSection section = kOptionSectionUnknown;
+    std::string title;
+    std::string argument;
+    std::unordered_map<std::string, std::string> opt_map;
+    std::string line;
+    // we only support single-lined statement.
+    while (
+        lf_reader.ReadLine(&line, Env::IO_TOTAL /* rate_limiter_priority */)) {
+      int line_num = static_cast<int>(lf_reader.GetLineNumber());
+      line = TrimAndRemoveComment(line);
+      if (line.empty()) {
+        continue;
       }
+      if (IsSection(line)) {
+        s = EndSection(config_options, section, title, argument, opt_map);
+        opt_map.clear();
+        if (!s.ok()) {
+          break;
+        }
 
-      // If the option file is not generated by a higher version, unknown
-      // option should only mean corruption.
-      if (config_options.ignore_unknown_options &&
-          section == kOptionSectionVersion) {
-        using VTuple = std::tuple<int, int, int>;
-        if (VTuple(db_version[0], db_version[1], db_version[2]) <=
-            VTuple(ROCKSDB_MAJOR, ROCKSDB_MINOR, ROCKSDB_PATCH)) {
-          config_options.ignore_unknown_options = false;
+        // If the option file is not generated by a higher version, unknown
+        // option should only mean corruption.
+        if (config_options.ignore_unknown_options &&
+            section == kOptionSectionVersion) {
+          using VTuple = std::tuple<int, int, int>;
+          if (VTuple(db_version[0], db_version[1], db_version[2]) <=
+              VTuple(ROCKSDB_MAJOR, ROCKSDB_MINOR, ROCKSDB_PATCH)) {
+            config_options.ignore_unknown_options = false;
+          }
         }
-      }
 
-      s = ParseSection(&section, &title, &argument, line, line_num);
-      if (!s.ok()) {
-        return s;
+        s = ParseSection(&section, &title, &argument, line, line_num);
+        if (!s.ok()) {
+          break;
+        }
+      } else {
+        std::string name;
+        std::string value;
+        s = ParseStatement(&name, &value, line, line_num);
+        if (!s.ok()) {
+          break;
+        }
+        opt_map.insert({name, value});
       }
-    } else {
-      std::string name;
-      std::string value;
-      s = ParseStatement(&name, &value, line, line_num);
-      if (!s.ok()) {
+    }
+    if (s.ok()) {
+      s = lf_reader.GetStatus();
+    }
+    if (s.ok()) {
+      s = EndSection(config_options, section, title, argument, opt_map);
+      opt_map.clear();
+    }
+    if (s.ok()) {
+      s = ValidityCheck();
+    }
```

Comment:
nit: `if (s.ok()) { break; }` after this validity check would be more readable to me then setting retry = false at the else in line 349. Or you could explicitly exit the loop on line 349.